### ISSUE TITLE
[css-color] [css-ui-4] Authors should have access to accent-color value to use in their code

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#5900]
Edit: [Resolved](https://github.com/w3c/csswg-drafts/issues/5900#issuecomment-2380321760) to do this

---

https://drafts.csswg.org/css-ui-4/#widget-accent

As discussed in today's telcon, there should be a keyword (akin `currentColor`) or other way to refer to the value of `accent-color` so that authors can use this in their designs as well.
The advantage of this over a custom property, is decoupling: that libraries can hook into existing code using `accent-color` (especially given proposed guidance to set it on the root), without the code needing to be aware of the existence of the library.

**Comment by bradkemper** [comment]
I agree, but I was thinking it should be an environment variable, e.g. `env(accent-color)`. 

Hopefully to do it in a way that doesn't allow using it for fingerprinting.

**Comment by samuelbradshaw** [comment]
I can see the value in both referencing the author-defined accent color on a section of the page, and referencing the user agent's default accent color. Why not both? :)

**Comment by SebastianZ** [comment]
If this should be made available via an environment variable, avoiding fingerprinting is an important point here. Access to `accent-color` probably needs to be restricted to avoid reading the actual value, otherwise the user agent, operating system or device could be determined when they use a unique color.

Though I guess that's something that applies to all environment variables.

Sebastian

**Comment by LeaVerou** [comment]
@SebastianZ This is something set in CSS, so I'm not sure how fingerprinting applies?

**Comment by SebastianZ** [comment]
Well, I assume @bradkemper's suggested environment variable will compute to an actual color value, which will be the color chosen by the user agent by default. So the environment variable needs to compute to the `auto` value or to itself and it needs to be ensured that there's no other way to get the actual color value via CSSOM or other means in order to avoid fingerprinting.

Sebastian

**Comment by frivoal** [comment]
I guess it can't be an env variable: such variables are evaluated independently of where you are in the tree, and thus would always return the UA's default accent color, which we cannot share for fingerprinting reasons (see in https://github.com/w3c/csswg-drafts/issues/6158). If we want the local accent-color, as set by the author, we need something context sensitive, so that wouldn't be an env variable, but something more like the `currentColor` keyword.

That said, I wonder if this is really worth the cost, as there's an easy work-around. If an author needs this, they can just as well set their own `--accent-color` custom property, and use its value to set both the actual `accent-color` property and whatever other place they wanted to use that color in.

**Comment by emilio** [comment]
So, fwiw. WebKit does expose the OS accent color via `-webkit-focus-ring-color`.

Gecko has `-moz-accent-color` for the Firefox UI but we don't expose it to websites. I think I agree with @frivoal that making a more magic keyword for this seems overkill.

**Comment by LeaVerou** [comment]
> I guess it can't be an env variable: such variables are evaluated independently of where you are in the tree, and thus would always return the UA's default accent color, which we cannot share for fingerprinting reasons (see in #6158). If we want the local accent-color, as set by the author, we need something context sensitive, so that wouldn't be an env variable, but something more like the `currentColor` keyword.

This brings up a more general point: Do we need something like `env()` that *is* context-sensitive?

> So, fwiw. WebKit does expose the OS accent color via `-webkit-focus-ring-color`.
> 
> Gecko has `-moz-accent-color` for the Firefox UI but we don't expose it to websites. I think I agree with @frivoal that making a more magic keyword for this seems overkill.

I think the fact that UAs are making up their own keywords for this indicates that there's a need for it.

**Comment by emilio** [comment]
Those two keywords are effectively prefixed system colors, not magic CSS values in any other way.

**Comment by samuelbradshaw** [comment]
Adding a system color like `NativeAccentColor` seems like an easy solution, but it doesn't solve the use case of custom third-party components, as mentioned in https://github.com/w3c/csswg-drafts/issues/5900#issuecomment-826843078. For components, the system accent color isn't what I want if a custom `accent-color` is defined.

**Comment by clshortfuse** [comment]
The new direction of Google's Material Design (Material You) will use one color to extrapolate a color scheme with multiple palettes. Unfortunately, we would need the RGB value (to get the hue-color-tone). Also, because it's central to the presentation of the page, it would best to be calculated before first paint. That means JS would run in `<head>`, calculate the palette, and get it all ready for the first paint.

I do worry about fingerprinting because of it though. I guess we can expose a permission? But that doesn't make sense the context of first-paint if the user has to accept. Right now I only care about web apps, so perhaps this value can be auto-granted permission on app install. That's probably a larger conversation though.

* https://github.com/material-foundation/material-color-utilities
* https://material.io/blog/science-of-color-design

**Comment by emilio** [comment]
@LeaVerou, let me know if you'd prefer me to open a different issue to expose the "native" accent color using a standard color pair (WebKit/Blink effectively expose them already with the non-standard `-webkit-focus-ring-color`) and keep this issue for the magic keyword.

I think these are sufficiently related but happy to split that out to a separate issue.

**Comment by LeaVerou** [comment]
Yes, I think these are separate issues.

**Comment by emilio** [comment]
Sure, filed https://github.com/w3c/csswg-drafts/issues/7347

**Comment by svgeesus** [comment]
Closed by https://github.com/w3c/csswg-drafts/commit/8c1fe16402cee71211530ef5283a0372c8740e7d

**Comment by emilio** [comment]
Not really, per https://github.com/w3c/csswg-drafts/issues/5900#issuecomment-1151240336 this is about a more magical mechanism that would compute to the current `accent-color` property value (a concrete proposal would be a `currentAccentColor` color, or something like that?).

But https://github.com/w3c/csswg-drafts/issues/7347 would solve some of the use cases raised here.

**Comment by jacobrask** [comment]
Adding some concrete use cases here in case that helps.

- we want the focus ring of a form input to match the accent-color, which can be different between different user journeys on the same website
- we want to reuse the accent-color for custom radio/checkbox/toggle components

We can't just set `:root { accent-color: var(--accent-color) }`, we'd need to explicitly set it on every element that uses `accent-color` (including within shadow dom), or always do `accent-color: hotpink; --accent-color: hotpink;` every time we want to toggle it on an individual form and not on an entire page.

For example, a checkout form might use one accent color, but a contact form on the footer on the same page should use a more neutral accent color.

**Comment by o-t-w** [comment]
This is a great idea. Most frontend developers are familiar with the `accent-color` property but on professional websites its more common to use `appearance: none` and customise the control, which limits the usefulness of `accent-color`. There are hundreds of component libraries out there. Currently they all use differently named custom properties. Imagine knowing how to style form controls without needing to look in the docs. Some websites use both native form controls and some custom form controls. With this property you could be confident you were styling all of them with a single line of CSS. I've built custom checkboxes/radios/etc plenty of times in various JavaScript frameworks over the years. Having a consistent way to style them is definitely worthwhile.

**Comment by josepharhar** [comment]
Supporting this use case with a color keyword sounds good to me.

In my mind, AccentColor/AccentColorText are about getting access to the OS system accent color rather than whatever accent-color was set to, but if people really want AccentColor/AccentColorText to behave that way then thats OK with me, especially since we havent shipped AccentColor/AccentColorText in chromium yet. If we create a new color keyword for this, then currentAccentColor sounds good to me.

**Comment by lukewarlow** [comment]
I would not want to overload the system keyword for this. I like currentAccentColor (scope creep can we get currentBackgroundColor too 🙏🏻 )

**Comment by josepharhar** [comment]
Lea mentioned offline that overloading the system keyword is better because of the case where the author doesn't set accent-color to anything. What would currentAccentColor resolve to when accent-color has its default value of "auto"?

**Comment by waterplea** [comment]
Without this feature `accent-color` is pretty useless, hopefully it gets added.

**Comment by clshortfuse** [comment]
Related https://github.com/chromium/chromium/commit/8a73948a2198aa9674ad806ce6c51cabc9baecfb

> With these changes enables the OS-defined accent color on Window to be used by:
> - HTML form elements that support CSS property-"accent-color" and set to use default system accent color
> - System color keyword - "AccentColor".

Tester - https://codepen.io/shortfuse/pen/LYgoKQe

**Comment by LeaVerou** [comment]
> I would not want to overload the system keyword for this. I like currentAccentColor (scope creep can we get currentBackgroundColor too 🙏🏻 )

Are there any use cases where authors need to access the OS default accent color *even* if there is an `accent-color` set? I don't see any, but I could well be missing them. But if there aren't, I don't understand the rationale for this — it complicates the language and the number of keywords authors need to learn with no discernible benefit. Remember theoretical purity is at the bottom of the PoC.

**Comment by clshortfuse** [comment]
> Are there any use cases where authors need to access the OS default accent color _even_ if there is an `accent-color` set? I don't see any, but I could well be missing them. But if there aren't, I don't understand the rationale for this — it complicates the language and the number of keywords authors need to learn with no discernible benefit. Remember theoretical purity is at the bottom of the PoC.

Personally, `accent-color` is useless for me. I have no interest in using system controls. I want to use custom elements and styled to match the system's accent's hue. That includes default, hover, active states. 

----

[Material Design](https://m3.material.io/styles/color/system/overview) has color schemes completely designed around one user accent color in RGB. From there you extrapolate the rest of the colors.  That's the default on Android, and if I want my Web Applications to look native, it starts there.

A live example of what the Web Components look like can be seen here: https://clshortfuse.github.io/materialdesignweb/components/buttons.html with the default based around a single color: #6750A4; the alt is based around #9a4522:

![image](https://github.com/user-attachments/assets/465f24db-c4c4-4bc9-a875-feea40f06026)
![image](https://github.com/user-attachments/assets/1e43f201-b9e7-4b58-b789-9e8ea3578e35)

---

For Windows 11 accents on the native controls also use the system color scheme:
![image](https://github.com/user-attachments/assets/ad122a7d-39c6-4780-b142-506f2b6e57fb)
![image](https://github.com/user-attachments/assets/35b122a7-4fe7-4be6-9ea3-b706081be0c1)
![image](https://github.com/user-attachments/assets/50ff9eff-e9ad-4f97-b5f6-733c19796ced)

If somebody where to want to build an Electron app that looks like native Windows then that color would be important.

----

iOS, iPadOS, Mac OS, tvOS all use variations of the system [accent color](https://developer.apple.com/documentation/swiftui/color/accentcolor) for controls. 

![image](https://github.com/user-attachments/assets/66e12ef1-586d-45a9-9f4a-0e82f4501340)
![image](https://github.com/user-attachments/assets/bb571580-7f8d-4351-9120-40fcbe6484b2)

---

Since we are talking Web Components, then you'd have to recreate the native looking controls as Web Components. Personally, I don't want to make one application in JS for Web, one in Kotlin for Android, one in Swift for iOS, and one in .NET for Windows. I use PWAs, which I feel is one of great strengths of Web. I've done the work for Android and ChromeOS (Material Design), and others like ([Ionic](https://ionicframework.com/docs/api/toggle)) have done the legwork for iOS. There are even frameworks like React Native that are built to extend that native feeling all the way to being actually native code on host OS's.

**Comment by samuelbradshaw** [comment]
> I have no interest in using system controls. …
> … I want my Web Applications to look native, …

Maybe I'm misunderstanding, but these sound like competing priorities. If you want it to look native, wouldn't you want to use native system controls?

**Comment by clshortfuse** [comment]
> > I have no interest in using system controls. …
> > … I want my Web Applications to look native, …
> 
> Maybe I'm misunderstanding, but these sound like competing priorities. If you want it to look native, why wouldn't you want to use native system controls?

Browsers don't style their controls to match native system controls. Either the native element doesn't exist as a HTML element (tooltip, navdrawer, switch, menu) or their styling is different (button, checkbox, text box, dialog). And I'm rather sure Chrome and Firefox have system-agnostic controls (same on every platform).

That difference is why frameworks like React Native and Ionic exist.

**Comment by kizu** [comment]
> Are there any use cases where authors need to access the OS default accent color even if there is an accent-color set? I don't see any, but I could well be missing them. But if there aren't, I don't understand the rationale for this — it complicates the language and the number of keywords authors need to learn with no discernible benefit. Remember theoretical purity is at the bottom of the PoC.

An additional point: even if there are such use cases, they will be possible by capturing the corresponding colors on the `:root` with registered custom properties, and accessing them later regardless of if they were overridden. It could be even possible to store both the light & dark versions of them via setting two different `color-scheme` values on `html` and `body`, and storing two sets of custom properties there.

So, I'm for the `accent-color` following the overridden value, as this will cover the more common use cases and will remove the current confusing behavior, and more complex cases will still be possible, even if a bit hacky.

**Comment by nt1m** [comment]
This illustrates @frivoal's concerns: https://jsfiddle.net/5mb7efou/

This is rendered like this in WebKit if you've picked the green accent color setting:

<img width="615" alt="image" src="https://github.com/user-attachments/assets/c4439ad5-acff-480c-9c47-8aee48f9fa55">

This is why florian doesn't think we should generally encourage users to use AccentColor in general.

**Comment by fantasai** [comment]
@nt1m I think we need to figure this out if we want to have a good default stylesheet for `appearance: base` form controls, though. There's no way for that stylesheet to use the value of `accent-color` if we don't have a keyword that references it.

**Comment by LeaVerou** [comment]
To summarize @frivoal’s concern: WebKit lies about `accentColor` but still uses the correct OS accent color in form controls. You can see this in action in this demo (set your OS default color to something else first): https://result.dabblet.com/gist/d9354b57012bc1f9eaf8edfbc2b39bed/30d00e3ac6eae4dab6e6ac5a05419c62975f5fd3

However:
1. This is a problem with the *existing* `accentColor` keyword, and if anything is *reduced* by this change. 
2. Even if the OS default were useless, authors having a universal "contract" to communicate accent color to components is still very valuable.
3. I’d argue this is a bug in WebKit and it should instead use the correct color and not expose it if they are so concerned about fingerprinting.

**Comment by clshortfuse** [comment]
Thanks for working out the minutiae for this issue. 

I believe that the acceptable contrast between AccentColorText and AccentColor is a "courtesy" from the UserAgent but in no means a strict expectation. If a developer wants 3:1, 4.5:1 or even 7:1 it appears as though this information may be computed as needed by the developer.

I'm still assuming AccentColor is a readable value and the matter of fingerprinting has yet to be finalized.

**Comment by svgeesus** [comment]
Trying to implement [the resolution](https://github.com/w3c/csswg-drafts/issues/5900#issuecomment-2380321760) and this issue is tagged `css-color-5` but that spec has no mention of `AccentColor`. It is [defined in CSS Color 4](https://drafts.csswg.org/css-color-4/#css-system-colors) and the [accent-color](https://drafts.csswg.org/css-ui-4/#propdef-accent-color) property is [defined in CSS Basic UI 4](https://drafts.csswg.org/css-ui-4/#widget-accent).

Which spec is supposed to be changed here? My guess is CSS UI 4.

@LeaVerou @frivoal

**Comment by LeaVerou** [comment]
My guess would be css-color (not a strong opinion though). The concept of `AccentColor` predates the `accent-color` property, and one could argue they can be implemented separately. The `accent-color` property is just one of many ways to set it.

**Comment by alisonmaher** [comment]
Spec updates added via https://github.com/w3c/csswg-drafts/pull/12733#event-19513127157
